lib/ukstore: Introduce the `ukstore` library (static)

craciunoiuc opened the pull request:
These commits introduce a new internal library to Unikraft, `ukstore`.
The role of this new library is to provide a standard way of passing
information getters/setters from one place to another.

The function pointers are stored in entries and entries are stored in
folders. These are structured in two different ways, statically and
dynamically. The static ones are used for information that does not
change at runtime.

To generate sections per library the AWK tool is used together with
the build system. This library has a dependency only to nolibc.

This PR does not depend on any other `ukstore` PR.
The old PR was marked as Draft.

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@gmail.com>

craciunoiuc committed:
lib/ukstore: Introduce library skeleton

The `ukstore` library has a dependency to `nolib`/`newlib` and to
`ukdebug`. As such they need to be selected to work.

The symbols exported will be provided by the `store.c` file.

The Makefile contains rules to build all files and generate files.
The Makefile first generates a list of all available libraries and
then generates two headers and a linker script to create and access
sections for each available library.

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@gmail.com>

craciunoiuc commented:
@skuenzer we will continue our discussion here. This can/should make it this release

I have 2 things I will comment upon because I'm not sure about them.
Also, please do a full pass over it, because I had to reorder the commits and I'm not 100% I removed everything dynamic-related (even though I checked it and it works)

also tagging: @marcrittinghaus @nderjung @mandrei12

craciunoiuc commented on lib/ukstore/store.c:
```diff
@@ -0,0 +1,1214 @@
+	}
+
+	case uk_store_get_entry_type(charp): {
+		char *to_set = malloc(sizeof(*val) * strlen(val));
+
+		if (unlikely(!to_set))
+			return -ENOMEM;
+
+		strcpy(to_set, val);
+
+		ret = e->set.charp(e->cookie, to_set);
+
+		free(to_set);
```
This makes sense, but doesn't feel right, we need to discuss how to manage memory inside it.
Most probably we need to pass to it also a buffer, or inform people that their buffer will be saved as a reference, not reallocated.

skuenzer replied:
I see what you mean, yeah... In this particular case I would document that the provided charp setter function of the library needs to take care of duplicating the given string. Here, I would just forward the argument as `const char *`. More tricky and less obvious is the `getter` that returns a charp.

craciunoiuc commented on lib/ukstore/store.c:
```diff
@@ -0,0 +1,1214 @@
+	case uk_store_get_entry_type(u8): {
+		__u8 val;
+
+		str = calloc(UK_STORE_NR_SIZE8_LEN, sizeof(char));
+		if (unlikely(!str))
+			return -ENOMEM;
```
Also here.

craciunoiuc commented on lib/ukstore/store.c:
```diff
@@ -0,0 +1,1214 @@
+	case uk_store_get_entry_type(s8): {
+		__s8 val;
+
+		str = calloc(UK_STORE_NR_SIZE8_LEN, sizeof(char));
+		if (unlikely(!str))
+			return -ENOMEM;
```
Also here.

craciunoiuc commented on lib/ukstore/store.c:
```diff
@@ -0,0 +1,1214 @@
+	case uk_store_get_entry_type(u16): {
+		__u16 val;
+
+		str = calloc(UK_STORE_NR_SIZE16_LEN, sizeof(char));
+		if (unlikely(!str))
+			return -ENOMEM;
```
Also here.

craciunoiuc commented on lib/ukstore/store.c:
```diff
@@ -0,0 +1,1214 @@
+	case uk_store_get_entry_type(s16): {
+		__s16 val;
+
+		str = calloc(UK_STORE_NR_SIZE16_LEN, sizeof(char));
+		if (unlikely(!str))
+			return -ENOMEM;
```
Also here.

craciunoiuc commented on lib/ukstore/store.c:
```diff
@@ -0,0 +1,1214 @@
+	case uk_store_get_entry_type(u32): {
+		__u32 val;
+
+		str = calloc(UK_STORE_NR_SIZE32_LEN, sizeof(char));
+		if (unlikely(!str))
+			return -ENOMEM;
```
Also here.

craciunoiuc commented on lib/ukstore/store.c:
```diff
@@ -0,0 +1,1214 @@
+	case uk_store_get_entry_type(s32): {
+		__s32 val;
+
+		str = calloc(UK_STORE_NR_SIZE32_LEN, sizeof(char));
+		if (unlikely(!str))
+			return -ENOMEM;
```
Also here.

craciunoiuc commented on lib/ukstore/store.c:
```diff
@@ -0,0 +1,1214 @@
+	case uk_store_get_entry_type(u64): {
+		__u64 val;
+
+		str = calloc(UK_STORE_NR_SIZE64_LEN, sizeof(char));
+		if (unlikely(!str))
+			return -ENOMEM;
```
Also here.

craciunoiuc commented on lib/ukstore/store.c:
```diff
@@ -0,0 +1,1214 @@
+	case uk_store_get_entry_type(s64): {
+		__s64 val;
+
+		str = calloc(UK_STORE_NR_SIZE64_LEN, sizeof(char));
+		if (unlikely(!str))
+			return -ENOMEM;
```
Also here.

craciunoiuc commented on lib/ukstore/store.c:
```diff
@@ -0,0 +1,1214 @@
+	case uk_store_get_entry_type(uptr): {
+		__uptr val;
+
+		str = calloc(UK_STORE_NR_SIZE64_LEN, sizeof(char));
+		if (unlikely(!str))
+			return -ENOMEM;
```
Also here.

craciunoiuc reviewed (commented):
Hey @skuenzer,

we need to discuss the allocations done here.
Basically, who does the allocation, the caller or the callee.

This is the case also for the `dynamic` part, but I'll do the changes also there when it's the case.

mandrei12 reviewed (commented):
I tested the static part and after checking the code I did not find anything wrong with it.

skuenzer commented on lib/ukstore/include/uk/store.h:
```diff
@@ -29,3 +29,344 @@
+
+/* All basic types that exist - use these when interacting with the API */
+enum uk_store_entry_type {
+	_LIB_UK_STORE___undef = 0,
```
You don't need to set a number for each enum if they are anyways linearly increasing. The C enumeration declaration will do this for you automatically. Just define the value for `undef`.

skuenzer replied:
For name the I would also suggest to add `ENTRY_TYPE` for better differentiation : `_LIB_UK_STORE_ENTRY_TYPE_<type>` (as inline with your `ENTRY_FLAG` definition).

skuenzer commented on lib/ukstore/exportsyms.uk:
```diff
@@ -0,0 +1,23 @@
+_uk_store_get_s64
+_uk_store_get_uptr
+_uk_store_get_static_entry
+_uk_store_free_entry
```
Put only the symbols here that you are actually introducing with this commit and the ones we need for static entries support.

craciunoiuc replied:
don't know how I missed them 😅

skuenzer commented on lib/ukstore/Makefile.uk:
```diff
@@ -29,6 +33,22 @@ $(LIBUKSTORE_BUILD)/uk_libs.in.new:
 		| tr ' ' '\n' > $@)
 
+$(LIBUKSTORE_GEN_INCLUDES_PATH)/uk/_store_libs.h: $(LIBUKSTORE_BASE)/gen_libs.h.awk $(LIBUKSTORE_BUILD)/uk_libs.in
+	$(call build_cmd,AWK,libukstore,$(notdir $@), \
```
Hey, I think the generation of id lists of the selected libraries should go into an own library: `libs/uklibid`. As it currently looks like, we will need something like this for the changes we work on `lib/ukdebug`. This is for something later: Could you provide a PR after we integrated this one that makes this split? While you do this, could you also adopt the rules to use the built-in `AWK` preprocessing rule that is available in the meantime. For example:
```Makefile
LIBUKSTORE_SRCS-y += $(LIBUKSTORE_BASE)/_store_libs.awk>.h
LIBUKSTORE_GEN_LIBS_AWKFLAGS    += -F '-'
LIBUKSTORE_GEN_LIBS_AWKINCLUDES += $(LIBUKSTORE_BASE)/uk_libs.in
```
...generates:
`(LIBUKSTORE_BUILD)/_store_libs.h`
from:
`$(LIBUKSTORE_BASE)/_store_libs.awk`
...as part of the `preprocess` build target. Note that the target name is derived from the source file name.

The purpose of this new build system extension is to make ther `Makefile.uk` more readable. Also, it can become really handy, for generating `.c` or `lds.S` files, the resulting files will be automatically included to the build process accordingly, e.g.,  `>.c` preprocessing target will cause also generating a C compile rule for this file.

craciunoiuc replied:
Yup, no problem, I can start working on it for the next release. (0.10)

Just so I understood clearly: Nothing to be done for this PR right?

skuenzer replied:
Yes 😄

skuenzer commented on lib/ukstore/gen_libs.h.awk:
```diff
@@ -0,0 +1,22 @@
+
```
Empty newline at the beginning of the file? 😄

craciunoiuc replied:
whoops, no checkpatch for awk 😄

skuenzer commented on lib/ukstore/include/uk/store.h:
```diff
@@ -29,3 +29,344 @@
+ * Adds an entry to the entry section of a library.
+ *
+ * @param entry the entry in the section
+ * @param e_type the type of the function
```
maybe add also example of entry types:
```
@param e_type Type for the entry, e.g., s8, u16, charp
```

skuenzer commented on lib/ukstore/include/uk/store.h:
```diff
@@ -29,3 +29,344 @@
+ *
+ * @param entry the entry in the section
+ * @param e_type the type of the function
+ * @param e_get getter pointer
```
I would also say that getter/setter are optional. Obviosly settin both to `NULL` at the same time would not make much sense, but it should not break anything if a user does it. So, no need to check arguments, just mention that `NULL` can be set, too as part of the parameter description.

skuenzer commented on lib/ukstore/include/uk/store.h:
```diff
@@ -29,3 +29,344 @@
+} __align8;
+
+/* Used internally for storage */
+struct uk_store_folder;
```
Introduce this later, when intorducing the dynamic entries

craciunoiuc replied:
yup, again, missed ;(

skuenzer commented on lib/ukstore/include/uk/store.h:
```diff
@@ -29,3 +29,344 @@
+ *
+ * @param entry the entry to free
+ */
+#define uk_store_free_entry(entry)	\
```
I think this is also for later, right... when we have dynamic entries?

skuenzer commented on lib/ukstore/include/uk/store.h:
```diff
@@ -29,3 +29,344 @@
+ * Searches for an entry in a folder in a library. Increases the refcount.
+ *
+ * @param libname the name of the library to search in
+ * @param foldername the name of the folder to search in
```
Add a comment that foldername is optional and reserved for future use with dynamic entries. Setting it to `NULL` makes sense

skuenzer commented on lib/ukstore/include/uk/store.h:
```diff
@@ -29,3 +29,344 @@
+ * Calls the getter to get it's value
+ *
+ * @param entry the entry to call the getter from
+ * @param outtype the type of the variable used
```
give an example of type: e.g., `s8, u8, uptr`

skuenzer commented on lib/ukstore/include/uk/store.h:
```diff
@@ -29,3 +29,344 @@
+ * Calls the set to set a new value value
+ *
+ * @param entry the entry to call the setter from
+ * @param intype the type of the variable used
```
I would also add here an example of types

skuenzer commented on lib/ukstore/store.c:
```diff
@@ -29,3 +29,1186 @@
+#include <uk/arch/types.h>
+#include <uk/store.h>
+
+struct uk_store_folder {
```
Can you only introduce the relevant structures and funtions for static entries? The others should be added with the dynamic entries patches.

skuenzer commented on lib/ukstore/include/uk/store.h:
```diff
@@ -29,3 +29,344 @@
+#define uk_store_libid_self()	\
+	uk_store_libid(__LIBNAME__)
+
+/* All basic types that exist - use these when interacting with the API */
```
I would add the list of simple types so that people can check what they have to enter when using the interfaces of this library.

craciunoiuc replied:
in the comment right? adding them now

skuenzer commented on lib/ukstore/Config.uk:
```diff
@@ -0,0 +1,5 @@
+config LIBUKSTORE
+	bool "ukstore: Run-time configuration and monitoring"
```
ukstore: Run-time configuring and monitoring

skuenzer commented on lib/ukstore/include/uk/store.h:
```diff
@@ -29,3 +29,344 @@
+};
+
+/* Transforms a simple data type to a ukstore specific one */
+#define uk_store_get_entry_type(simple_type)	\
```
Hum...  🤔 this could be a confusing name for the macro because it is close to `uk_store_entry_get_type()`, what do you think? What if we just name it `UK_STORE_ENTRY_TYPE()` because it is a pure pre-processor macro and not a getter function. 🤔 This would be inline with the macro `UK_STORE_ENTRY_ISSTATIC`, for example.

craciunoiuc replied:
yup, makes sense

skuenzer commented on lib/ukstore/store.c:
```diff
@@ -29,3 +29,1186 @@
+	__atomic refcount;
+} __align8;
+
+#define UK_STORE_NR_MAX_LEN 32
```
I would suggest a different name. It wasn't clear to me what these definitions mean until I saw the code using them. Additionally, they are also internal only, so you do not need the `UK_STORE` prefix.
What would you think of:
```c
#define _U8_STRLEN (3 + 1) /* biggest num (127) has 3 digits plus terminating '\0' */
#define _S8_STRLEN (1 + _U8_STRLEN) /* space for potential sign in front */
#define _U16_STRLEN (5 + 1)
#define _S16_STRLEN (1 + _U16_STRLEN)
#define _U32_STRLEN (10 + 1)
#define _S32_STRLEN (1 + _U32_STRLEN)
#define _U64_STRLEN (19 + 1)
#define _S64_STRLEN (1 + _U64_STRLEN)
#define _UPTR_STRLEN (2 + 16 + 1) /* this is for uptr (64bit) in hex with space for leading `0x` prefix */
```

skuenzer commented on lib/ukstore/store.c:
```diff
@@ -29,3 +29,1186 @@
+}
+
+int
+_uk_store_get_charp(const struct uk_store_entry *e, char **out)
```
For your question, I think we should document that the caller must free the string and then it is fine. The library provided charp getter function would also allocate a string for an output with a total custom size. I think it would be fine to tell the API user that strings need to be free'd.
We can provide another variant, like `_uk_store_get_ncharp(const struct uk_store_entry *e, char *out, size_t maxlen)`. Here, you handover a buffer to write to and its size. Things would then be truncated. You can use `snprintf()` within the implementation. In the charp-to-charp case you would need to do `snprintf(out, maxlen, "%s", val)`.

craciunoiuc replied:
I'm okay with `_uk_store_get_ncharp`, but, just so you know, that would change the define macro from:

```c
#define uk_store_get_value(entry, outtype, out)	\
	_uk_store_get_ ## outtype((entry), (out))
```

to something like

```c
#define uk_store_get_value(entry, outtype, out, maxlen)	\
	_uk_store_get_ ## outtype((entry), (out), (maxlen))
```

and this would change the signatures too. Is this ok, should I go for it?

skuenzer replied:
Hum... I would actually treat the ncharp as an additional variant and keep what you have. maxlen wouldn't be usable for the other outtypes. The thing is also to define maxlen for getters that return charp already. Because our interface does imply any size limitations for charp it is likely that the output will be truncated with the ncharp varaint.

craciunoiuc replied:
ok, I did it separately, like this

```c
#define uk_store_get_value_ncharp(entry, out, maxlen)	\
	_uk_store_get_ncharp((entry), (out), (maxlen))
```

\+ adding one new typedef/elem to the union/enum item of course

skuenzer commented on lib/ukstore/store.c:
```diff
@@ -29,3 +29,1186 @@
+		ret = e->get.u8(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		sprintf(str, "%" __PRIu8, val);
```
if you use `snprintf` instead, the string will be `\0`-terminated in the target buffer - even if truncated but you size them big enough. You will also stay within bounds and you could just `malloc` the string buffer instead of `calloc`.

craciunoiuc replied:
I'll change all of them now

skuenzer commented on lib/ukstore/store.c:
```diff
@@ -29,3 +29,1186 @@
+		ret = e->get.uptr(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		sprintf(str, "%p", (void *) val);
```
I would suggest to use `"0x"__PRIx64` as format. `%p` is sometimes differently implementented across libcs. I have seen it once without leading `0x` prefix and once with.

skuenzer replied:
Actually, double-check that we use the same expected format when we have a setter and get a charp as input. Ideally, the function should support receiving hex with `0x` prefix and without.

craciunoiuc replied:
I'll just do it like this for that case

```c
		if (strncmp(val, "0x", 2) == 0)
			ret = sscanf(val, "0x%" __SCNx64, (void **) &to_set);
		else
			ret = sscanf(val, "%" __SCNx64, (void **) &to_set);
```

skuenzer reviewed (changes requested):
Hey @craciunoiuc,
thanks for the separation of the PR. I have only a few comments to this PR, mostly style related. Please also move pieces (functions, symbols, etc.) that belong to dynamic entries to the dynamic entry PR.
Let me know when you are ready. In fact I would like to include this PR (static entries) to this release.
Thanks, Simon

mandrei12 reviewed (commented):
I retested the static part and I did not find anything wrong with it.

craciunoiuc commented:
Hey @skuenzer,

Ready for round 2? (more like round N 😄)

skuenzer commented on lib/ukstore/include/uk/store.h:
```diff
@@ -29,3 +29,350 @@
+ * @return 0 or error on fail
+ */
+#define uk_store_get_value_ncharp(entry, out, maxlen)	\
+	_uk_store_get_ncharp((entry), (out), (maxlen))
```
I wouldnt put the macro in between for the ncharp type. Declare the symbol directly.

craciunoiuc replied:
Sure

I wanted to keep it uniform, with the rest

skuenzer commented on lib/ukstore/include/uk/store.h:
```diff
@@ -29,3 +29,350 @@
+ * @param out space to store the result of the call
+ * @return 0 or error on fail
+ */
+#define uk_store_get_value(entry, outtype, out)	\
```
This may fail when you specify ncharp as type because of the missing maxlen argument. I would actually exclude ncharp from the macro resolution. See my comment below....

skuenzer commented on lib/ukstore/store.c:
```diff
@@ -29,3 +29,1267 @@
+		return ret;
+	}
+
+	case UK_STORE_ENTRY_TYPE(charp): {
```
If you really add ncharp as entry type you need to handle it here too, but actually - I would exclude ncharp from the entry type and just have this one additional interface symbol.

craciunoiuc replied:
it will just fall in the default right? (if it's either undefined or something else)

skuenzer replied:
We have a Union so just one function type can be registered. If we call it with the wrong casting, it will fail.

skuenzer commented on lib/ukstore/include/uk/store.h:
```diff
@@ -29,3 +29,350 @@
+		uk_store_get_u64_func_t    u64;
+		uk_store_get_uptr_func_t   uptr;
+		uk_store_get_charp_func_t  charp;
+		uk_store_get_ncharp_func_t ncharp;
```
I wouldnt add the extra ncharp case currently because you will need to handle it in all switch cases as well.

skuenzer commented on lib/ukstore/store.c:
```diff
@@ -29,3 +29,1267 @@
+
+	case UK_STORE_ENTRY_TYPE(charp): {
+
+		ret = e->get.ncharp(e->cookie, out, maxlen);
```
Hum :thinking:. The entry is type of `charp` but a `ncharp` callback is called? I was assuming that we take the `charp` interface, do an `snprintf()` and then free the allocated buffer before returning this function. Of course, this is not zero-copy anymore but hopefully the `charp` interface is most rarely used.

craciunoiuc replied:
makes sense, especially if we are removing `get.ncharp` anyway

skuenzer commented on lib/ukstore/store.c:
```diff
@@ -29,3 +29,1267 @@
+	int ret;
+
+	UK_ASSERT(e);
+
```
In all the getter variabts and setter variants you should make sure that a getter is available. Otherwise return `-ENOENT`. Since get is a union you can do a non-NULL check of just one preferred field. Please add this to all getter and setter API functions.

craciunoiuc replied:
I can just check if the enum is `_LIB_UK_STORE_ENTRY_TYPE___undef`  -  I will do it like this

craciunoiuc replied:
```
	if (unlikely(e->type == UK_STORE_ENTRY_TYPE(undef)))
		return -ENOENT;
```

craciunoiuc replied:
But doesn't it already go in the default if the type is something other than the ones used?

```
	default:
		return -EINVAL;
	}
```

craciunoiuc replied:
I added the check anyway, hope it's okay

skuenzer replied:
It is for the case that we have get-only or set-only entries. Actually I would return -EIO if you set on a get-only entry (and vice versa). I think EIO is more appropriate for this case

skuenzer commented on lib/ukstore/store.c:
```diff
@@ -29,3 +29,1267 @@
+
+
+int
+_uk_store_get_ncharp(const struct uk_store_entry *e, char **out, size_t maxlen)
```
since we do not chnage the allocation of the `out` buffer, a single reference is good enough:
```c
int
uk_store_get_ncharp(const struct uk_store_entry *e, char *out, size_t maxlen)
```

skuenzer commented on lib/ukstore/store.c:
```diff
@@ -29,3 +29,1267 @@
+{
+	int ret;
+
+	UK_ASSERT(e);
```
assert that out exists, e.g.:
`UK_ASSERT(out)`

skuenzer reviewed (commented):
Can you add the get-only/set-only check?

mandrei12 reviewed (approved):
I retested the static part and did not find anything wrong with it.

Reviewed-by: Andrei Mutu <andrei.mutu@stud.acs.upb.ro>

skuenzer commented on lib/ukstore/include/uk/store.h:
```diff
@@ -29,3 +29,345 @@
+enum uk_store_entry_type;
+struct uk_store_entry;
+
+#define __UK_STORE_STATIC_ENTRY(entry, lib_str, e_type, e_get, e_set)	\
```
hum... did you try compiling instrumented code when the library is off. Because there is no definition of the struct and datatypes, I guess this registration will fail. I remember that we added this in order to keep type checking although it is not used. In fact - we may want to fix this by making the struct and the simple types always available.

craciunoiuc replied:
I don't know how this got here 😅 I'm removing it now

skuenzer commented on lib/ukstore/include/uk/store.h:
```diff
@@ -29,3 +29,345 @@
+#ifndef __UK_STORE_H__
+#define __UK_STORE_H__
+
+#include <string.h>
```
When the library is not part of the built, this header that is always available should not introduce unnecessary dependencies. Your library can't express any if not part of a build.

skuenzer replied:
So I would move header inclusion to line 97.

skuenzer replied:
When moving, make sure that headers that are required for definitions wheen the library is off, stay here. I hope these are only general unikraft headers (`/include`) and/or libc provided headers.

craciunoiuc replied:
There are no library dependencies when the library is off, it's fine

skuenzer commented on lib/ukstore/include/uk/store.h:
```diff
@@ -29,3 +29,345 @@
+} __align8;
+
+/* Flags if an entry is static or dynamic */
+#define UK_STORE_ENTRY_FLAG_STATIC	(1 << 0)
```
in order to avoid duplication, put common defintions (when library is on or off) before the guard checking if the library is part of the built: line 48

craciunoiuc replied:
I think there is nothing that is common between the two if/else cases. I'll just move the `#endif` lower.

skuenzer commented on lib/ukstore/include/uk/store.h:
```diff
@@ -29,3 +29,345 @@
+	((entry)->flags & UK_STORE_ENTRY_FLAG_STATIC)
+
+#define uk_store_get_entry(libname, foldername, entryname)	\
+	do { } while (0)
```
But, actually, the more I think about this, we should not provide any replacements for query calls. If someone wants to retrieve or set something from/to `ukstore`, that code has to make sure that the dependency to `lib/ukstore` is fulfilled. It is just convenience if we can transparently comment out with the preprocessor entry registration from libraries as soon as `ukstore` is not part of a built.

craciunoiuc replied:
Ok, I'm deleting everything except the entry registration

skuenzer reviewed (changes requested):
Hey @craciunoiuc , I just found now that we have some problems with compiling instrumented code when the library is not selected as part of a build. Can you double-check? Sorry that I did not see this earlier.
Simon

mandrei12 commented:
It still works after the changes.

craciunoiuc committed:
lib/ukstore: Add static entry registration

This commit introduces a method to allocate static elements in library
sections. Each library has its own section generated at build.
This type of entries do not take up time and are not modified.

These will be used to pass system information getters/setters
throughout the system in an easy-enough manner.

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@gmail.com>

skuenzer reviewed (approved):
Thanks a lot @craciunoiuc for the quick changes! This looks good.

Reviewed-by: Simon Kuenzer <simon.kuenzer@neclab.eu>
Approved-by: Simon Kuenzer <simon.kuenzer@neclab.eu>
